velox/exec/Task.cpp

```diff
@@ -418,6 +420,51 @@ void Task::removeSpillDirectoryIfExists() {
   }
 }
 
+/// Builds the query trace config.
+std::optional<QueryTraceConfig> Task::maybeMakeTraceConfig() const {
+  const auto& queryConfig = queryCtx_->queryConfig();
+  if (!queryConfig.queryTraceEnabled()) {
+    return std::nullopt;
+  }
+
+  const auto queryTraceDir = queryConfig.queryTraceDir();
+  VELOX_CHECK(
+      !queryTraceDir.empty(),
+      "Query trace enabled but the trace dir is not set properly");
+
+  const auto queryTraceNodes = queryConfig.queryTraceNodes();
+  if (queryTraceNodes.empty()) {
+    LOG(WARNING) << "Query trace enabled but no target nodes.";
+    return std::nullopt;
+  }
+
+  std::vector<std::string> nodes;
+  folly::split(',', queryTraceNodes, nodes);
+  const std::unordered_set<std::string> nodeSet(nodes.begin(), nodes.end());
+  VELOX_CHECK_EQ(nodeSet.size(), nodes.size());
+
+  const auto queryTraceWriteBufferSize = queryConfig.queryWriteBufferSize();
+
+  return QueryTraceConfig(nodeSet, queryTraceDir, queryTraceWriteBufferSize);
+}
+
+std::string Task::createTraceDirectory(const std::string& subDir) const {
+  std::lock_guard<std::mutex> l(dirCreateMutex_);
```

Comment:
Do we expect concurrent call to the same directory? If not, then we don't need the lock. I guess if it could happen if there are multiple operators need to be traced from the same driver? If it is the case, we might need to check if directory already exists here and return with no-op if it exists. Note velox fs mkdir might fail with it already exists.